Rewrite the plugin

donnchawp opened the pull request:
Initial import of framework for rewrite.
Also includes some work on configuration saving.

stodorovic committed:
Improve shutdown callback (#367)

* Improve shutdown callback

* Minor corrections, check is buffer empty

* Update wp-cache-phase2.php

* Add is_404 into wp_super_cache_query_vars

* Added wpsc_only_cache_known_pages filter

The wpsc_only_cache_known_pages filter returns 1  so that unknown page types are not cached. If the filter returns 0 even unknown page types will be cached, if all other conditions for caching are satisfied.

* Add method_exists as additional protection

donnchawp committed:
Give feeds on blogs with the plugin inactive a short TTL (#366)

* Give feeds on blogs with the plugin inactive a short TTL

Some servers have several networks running on one install and the server
admin may forget to activate the plugin on each network. Or a network
admin may disable the plugin for some reason on one blog, but caching
will still occur because it's loaded by WordPress because the WP_CACHE
constant is defined.
Unfortunately this stops wp-cache.php loading where wp_cache_setting()
is defined. That function is used to write a timestamp to the config
file whenever a post is updated. When the plugin is inactive, the feeds
will be given a TTL of 60 seconds so they do not become stale.
ref:
https://wordpress.org/support/topic/fatal-error-while-updating-post-or-publishing-new-one/

* If function DOESN'T exist, then short ttl

Reverse this logic. If the function doesn't exist then the plugin is not
activated and then set the short TTL.

donnchawp committed:
Remove last line from the file. (#369)

donnchawp committed:
Bump version to 1.5.5 and update changelog (#370)

stodorovic committed:
Minor corrections for debug and replace spaces with tabs (#371)

donnchawp committed:
Only realpath() the directory as it doesn't work on files. (#377)

Fixes #376

donnchawp committed:
Don't set this as it's not always saved when this function is called. (#379)

This function is called when the debug log is toggled, so setting this
here can confuse users who think that cache comments are active. It's
set further down in the function.

donnchawp committed:
Add plugins page to the API. (#382)

* Add plugins page to the API.

Update doesn't work yet.

* REST API: Remove html from desc and add url field to simplify.

Also run description and title through strip_tags() so custom plugins
don't send html to REST API clients.

* Move plugins settings to /plugins/ endpoint and add update code

* Force this plugin to use 0/1 in settings rather than yes/no strings.

Fixed so the plugin can be toggled through the REST API.
https://github.com/Automattic/wp-calypso/pull/18002#issuecomment-333305992

mackensen committed:
Test realpath against realpath; fixes #380 (#381)

Apply realpath() to get_supercache_dir() when checking $path as that has already been run through realpath().

ockham committed:
Plugins: Some copy fixing (#393)

Also, update docs link, and use a placeholder for it.

donnchawp committed:
Don't echo sensitive info to the browser. Props @janrenn. (#394)

Fixes #388

mathieuhays committed:
add wpsc_enable_wp_config_edit filter (#392)

* add wpsc_enable_wp_config_edit filter #372

* Add debug log when config file cannot be updated.

Warn the user through the debug log if the global config file can't be updated.
Also fix the indenting in part of the patch.

* Check if wp_cache_debug() exists

The function wp_cache_debug() won't exist when the plugin is activated so remove that call, and add a function_exists() around the one in the disable plugin function.

* separate wp-config file permission check

stodorovic committed:
Fixes PHP notices related to comments (#386)

* Fixes PHP notices related to comments

* Improve detection of delete_all in wp_cache_get_postid_from_comment

donnchawp committed:
Replace spaces by tabs (#395)

donnchawp committed:
Bump version to 1.5.6 and update changelog (#396)

donnchawp committed:
Check that wp_cache_setting exists before using it. (#399)

Fixes #398
Plugins are loaded really early so not everything is loaded including
this function.

donnchawp committed:
Bump to 1.5.7 and update changelog (#400)

donnchawp committed:
Update feeds when site updates (#403)

Store a record of archive feed pages that can be deleted when the cache
is updated. This will help keep feeds fresh. Do not record feeds for
single pages. A maximum of 50 feeds is recorded to avoid bloating the
options table.
Will avoid writing to the config file and causing race issues reading
and writing.

donnchawp committed:
Be extra paranoid. Check the location by using prune_super_cache() (#404)

The function prune_super_cache() checks the location of the file about
to be deleted so it's safer to use than unlink.

donnchawp committed:
Check the query var is_feed too when caching feed locations (#406)

I hardly ever see the correct content-type but when it's set then
$is_feed won't be set so check the cached query var to store the feed in
the feed_list.

donnchawp committed:
Don't preload posts with rejected strings (#407)

ref: https://wordpress.org/support/topic/wp-super-cache-ninja-forms/

donnchawp committed:
Cast as an array to avoid warnings

Somehow this gets set to null:
https://wordpress.org/support/topic/php-warning-in_array-expects-parameter-2-to-wp-cache-phase2-php-on-line-436/

donnchawp committed:
If the http host is empty get the path from the permalink (#408)

This avoids a PHP warning when using WP from the command line.

donnchawp committed:
Add/1 5 7 1 (#409)

Bump version to 1.5.7.1 and update changelog

donnchawp committed:
Bump tested version to 4.8.2 (#410)

Commit:
Use $_SERVER[ 'SERVER_PROTOCOL' ] to determine correct HTTP protocol version for response headers (#412)

See: http://php.net/manual/en/function.header.php#92305

Commit:
Use $_SERVER[ 'SERVER_PROTOCOL' ] to determine correct HTTP protocol version for response headers

See: http://php.net/manual/en/function.header.php#92305

(Last change I needed to propose.)

Commit:
Use double quotes.

Woops.

alexsanford committed:
Add more escaping to CDN settings page

The additional escaping is pretty paranoid, but should increase the page
security significantly.

donnchawp committed:
Add a log viewer to filter unwanted requests from debug log (#378)

* Add a log viewer to filter unwanted requests from debug log

* Move wpsc_create_debug_log() to wp-cache-phase1.php

Moving this function allows the debug log to be created correctly if
it's deleted and a function in the phase1 file logs something.

donnchawp committed:
Make writing to the config file more robust. (#402)

* Make writing to the config file more robust.

If multiple writers are interacting with the config file at one time
they may corrupt the file so write to a tmp file first and rename.
Also attempt to load the file until successful.

* Add some whitespace for readability

donnchawp committed:
Replace link and add warning about using "777" permission. (#421)

Fixes #419

donnchawp committed:
If a CPT has no rewrite rule then don't clear the cache. (#420)

Fixes #45
If a custom post type has no rewrite permalink then the plugin will
clear the entire cache in wp_cache_post_id_gc() because
get_current_url_supercache_dir() will strip the ?x=y from the URL,
resulting in the URL being the homepage. The plugin then deletes
everything below that.
See also https://github.com/Automattic/wp-super-cache/pull/66

stodorovic committed:
Add redirects/rest api/is_robots into wp-cache-phase2.php (#373)

* Add redirects/rest api into wp-cache-phase2.php

* Check 304, optimize conditions.

* Use hook status_header, more optimizations.

* Add wc-api, optimize sitemap detection

* Improve code by using is_robots, fixes PHP notice

jom committed:
Support all callables in `do_cacheaction()` (#425)

Fixes #418

Jon007 committed:
fixes #422 make known users anonymous / wp_cache_make_known_anon does not check DOING_AJAX (#423)

fixes #422 make known users anonymous / wp_cache_make_known_anon does
not check DOING_AJAX

donnchawp committed:
Move this function so it's available when wp-cache.php is not loaded (#426)

If a user activates the plugin on one blog of a multisite to admin other
blogs then wp-cache.php won't be loaded on those other blogs.

stodorovic committed:
Use get_term_link, get_permalink instead slug of taxonomy and post_name (#389)

Introduce function wpsc_delete_post_archives. It's replacement for wpsc_delete_cats_tags. It includes more dependent archive pages (custom taxonomies, post_type archive, author archive).
Use get_term_link, get_post_type_archive_link, get_author_posts_url. Better compatibility with other plugins.
Fixes #387

donnchawp committed:
If a URL has the character "?" then don't delete it. (#427)

donnchawp committed:
If wpsc_deep_replace() is missing then stop with an error message. (#428)

Fixes #75
Some users have unusual server configurations that means
advanced-cache.php and wp-cache-phase1.php aren't loaded when the
WordPress dashboard is running. This code exits with an error message if
that happens because the advanced settings page cannot save any updates.

stodorovic committed:
Improve PHP 7.0 compatibility (#429)

donnchawp committed:
Register hook in is_admin() to avoid unnecessary queries. props @krzyc. (#430)

ref:
https://wordpress.org/support/topic/register_uninstall_hook-should-only-be-called-when-is_admin-is-true/

donnchawp committed:
Fix/cache comments default on (#433)

* Cache comments are displayed by default so show that in settings page.

* Save cache comments status

* Added $wp_super_cache_comments to config file.

donnchawp committed:
Only send preload restart emails once a day. (#432)

* Only send preload restart emails once a day.

Ref:
https://wordpress.org/support/topic/preload-may-have-stalled-spammy-24-7-reports/
Shows an admin notice if the email isn't sent.

* Only show warning on wpsupercache settings page

donnchawp committed:
Check that needle is not empty before searching for it. (#437)

If the CDN setting "Exclude if substring" is blank a warning will be
generated when you look for it. This fixes that and uses stripos instead
of stristr as it's faster.
ref: https://wordpress.org/support/topic/another-empty-needle-issue/

donnchawp committed:
Fix some minor PHP warnings. (#438)

Ref: https://wordpress.org/support/topic/undefined-index-98/

donnchawp committed:
Don't annoy Nginx users so much with mod_rewrite warnings. (#434)

* Don't annoy Nginx users so much with mod_rewrite warnings.
Fixes #30
Nginx doesn't use mod_rewrite so don't annoy them about updating rules.

* Add link to Nginx rules if web server detected.

donnchawp committed:
Fix/update changelog bump 158 (#435)

* Bump version to 1.5.8 and update changelog

* Another changelog entry

* Bump test verion to 4.8.3 and update changelog

Commit:
Use SERVER_PROTOCOL to determine HTTP protocol (#439)

Somehow this change got overwritten many several commits ago.

timse201 committed:
link updates (#442)

Change http -> https links, and update old links to new.

donnchawp committed:
Move wp_cache_setting to phase1 file, and load everything on activation (#443)

* Move wp_cache_setting to phase1 file, and load everything on activation

The function wp_cache_setting() is used by the debug logger in
wp-cache-phase1.php so we need to define that function earlier.
The activation script then has to include the phase1 script so the
plugin can record settings.

* Move is_writeable_ACLSafe to phase1 too as it's used there.

donnchawp committed:
Make this more informative by including the $dir. (#448)

donnchawp committed:
Invalidate the config file opcache when it's updated. (#444)

Ref:
https://wordpress.org/support/topic/you-must-set-wp_cache-and-wpcachehome-in-your-wp-config-php-for-this-plugin/

donnchawp committed:
Fix/tidy up static page cache admin (#449)

* Invalidate the config file opcache when it's updated.

Ref:
https://wordpress.org/support/topic/you-must-set-wp_cache-and-wpcachehome-in-your-wp-config-php-for-this-plugin/

* Move this up to tidy up code when deleting a static front page.

timse201 committed:
echo __ -> _e and more clear button description (#446)

timse201 committed:
license uri added (#451)

Add license URL

donnchawp committed:
Move functions in phase1 file into phase2 file and load that everywhere (#465)

This will simplify loading the plugin as we don't have to worry about
functions existing before we use them.

stodorovic committed:
Cleanup some outdated code (#458)

Ismail-elkorchi committed:
Fix PHP Coding Standards errors (#463)

* Fix wpcs : awaitingmoderation.php

* Fix wpcs : badbehaviour.php

* Fix wpcs : domain-mapping.php

* Fix wpcs : dynamic-cache-test.php

* Fix wpcs : jetpack.php

* Fix wpcs : multisite.php

* Fix wpcs : searchengine.php

* Fix wpcs : wptouch.php

* Fix some errors

* Process of feedback

* Fix very minor whitespace typo

Add a space where none was before.

donnchawp committed:
Fix minor typo (#467)

stodorovic committed:
Little optimization - rest update settings (#460)

stodorovic committed:
Fix isset typo (#468)

timse201 committed:
readme overwork (#452)

* github styles and restructure

* additional styles

* Update readme.txt

* Rename readme.txt to readme.md

* header

* header

* development and documentation section

* deleted some unnecessary points

* headings

* wp 4.9

* Added link to translation page

* Add back warning about temporary nature of cache.

The cache directory is for temporary files only. Users have been caught out by this in the past so even though it's stating the obvious, this warning has to go in.

donnchawp committed:
Rename this back as it's needed for WordPress.org (#471)

donnchawp committed:
Use example.com as an example as it's safe to use this way. (#472)

donnchawp committed:
Add www redirect rules to readme.txt FAQ (#474)

Fixes #455

donnchawp committed:
Make explanation of preload mode clearer. (#477)

In response to
https://wordpress.org/support/topic/preload-and-garbage-collection-2/

stodorovic committed:
Change Tested up to and add Requires PHP (#476)

donnchawp committed:
Fix/dynamic cache test (#478)

* Explain how to hide secret tag. Props @Dravon5X
See #470
The secret tag used in this plugin can be viewed by site visitors if
they load up the cached page directly. Here's a few ways to stop that
happening.

* call_user_func_array() which we just started to use expects a value
Fixes 470

donnchawp committed:
Revert most of #446 as it caused problems displaying html (#479)

Fixes #459

stodorovic committed:
Fixes PHP notices (#383) and improve detects of backend request - rev (#475)

* Fixes PHP notices (#383) and improve detects of backend request - revision 2

* Check if get_option exists

Commit:
PHPCS fixes for ossdl-cdn.php (#486)

* PHPCS fixes for ossdl-cdn.php

* Update ossdl-cdn.php

Commit:
PHPCS fixes for advanced-cache.php (#482)

Commit:
PHPCS fixes for wp-cache-base.php (#483)

* PHPCS fixes for wp-cache-base.php

* Remove trailing whitespace

Minor edit to remove trailing whitespace.

Commit:
PHPCS fixes for wp-cache-phase1.php (#485)

* PHPCS fixes for wp-cache-phase1.php

* Minor edit to remove trailing whitespace.

donnchawp committed:
Dynamic caching must be enabled to use dynamic caching. (#491)

Fixes #490
If dynamic caching isn't enabled on the advanced settings page then the
set tag will be visible on the first load of a page.

stodorovic committed:
Little cleanup 1 (#488)

Fix white spaces/docblock after last few PRs.
Fix some names, remove unused global variables ( ossdl-cdn.php ).
Remove php closing tag from the end of plugins files ( awaitingmoderation.php, dynamic-cache-test.php, searchengine.php, badbehaviour.php, jetpack.php, domain-mapping.php, multisite.php, wptouch.php ).
Remove else to reduce complexity (recommend by phpmd and codeclimate).
Improve conditionals for checking wpnonce.

donnchawp committed:
Add note about mod_pagespeed in readme.txt (#492)

This Apache module may interfere with the cache tester. Ref:
https://wordpress.org/support/topic/cache-tester/

donnchawp committed:
Bump version to 1.5.9 and update changelog (#497)

stodorovic committed:
Fix issues in multisite plugin (#501)

stodorovic committed:
Fixes wp-cli plugin deactivate/activate (#499)

stodorovic committed:
Little cleanup 2 (#495)

donnchawp committed:
$htaccess_path defines the path to the global .htacess file. (#507)

* $htaccess_path defines the path to the global .htacess file.
Fixes  #350
Sometimes it's impossible to be flexible enough to cover every sort of
hosting that's out there. Set $htaccess_path to the path of the
.htaccess file to fix operations that write and read from that file.
Set that variable in wp-config.php or wp-cache-config.php

* Add troubleshooting docs on using $htacces_path

stodorovic committed:
Fix 'cannot redeclare gzip_accepted()' (#511)

KY4U committed:
Correct the renaming of tmp_wpcache_filename (removed unnecessary slash in path)  which caused renaming to fail. (#516)

Sentynel committed:
Add check for Jetpack mobile theme cookie (#515)

stodorovic committed:
Optimize wp_cache_phase2 and create wpsc_register_post_hooks (#508)

Includes fixes from #506

domtra committed:
display correct WPCACHEHOME path (#513)

The WPCACHEHOME constant needs a trailing slash.

stodorovic committed:
Optimize cache enable/disable and update_mod_rewrite_rules (#500)

* Don't show htaccess form if cache is disabled

* Prevent testing page if  is false

* Optimize cache enable/disable functions

clementduncan committed:
Post Update now clears category cache (#519)

If a post was updated, it would clear the homepage and post cache, but the category pages, i.e. category/stories/... wouldn't clear. The function was set to only update categories when a post was changed from un-published to published or vice versa.
It's been changed so it clears the category cache when a post goes from un-published to published or a change is made when it's published.

donnchawp committed:
Various fixes for saving the debug page form (#542)

* Various fixes for saving the debug page form
Fixes #541
Added wpsc_debug_username() to set $wp_cache_debug_username
Fixes missing $key parameter when setting globals in REST API update
Check if a GLOBALS variable was set before using it.
Check if wp_cache_debug_ip and wp_cache_debug_ip are defined before
using them.

* set_global() takes two parameters rather than one.

* Deal with input from the REST API and wp-admin as they're slightly
different

rmpel committed:
expert-caching and empty parameters, like ?amp, should not serve cached page (#533)

zigasancin committed:
Tiny Yslow description fix (#527)

The Firebug add-on doesn't exist anymore, nor does the extension Yslow for it, so it's time for a tiny wording change in the Yslow description.

bantilan committed:
Fixed iPad user agent conflict with wp_is_mobile() (#525)

Adding iPad user agent in the mobile_browser, as it conflict's with the wp_is_mobile() where iPad detected as mobile. This lead to caching a mobile version of a page when first loaded to iPad and cause displaying a mobile version for desktop.

donnchawp committed:
Hide opcache_invalidate() warnings since it's disabled some places. (#543)

Fixes #536
Those hosts may run into problems with the REST API since the config
file will be reloaded with old settings.

donnchawp committed:
Check that HTTP_REFERER exists before checking it. Props Jon007. (#544)

Fixes #538

donnchawp committed:
Replace Cron View" with WP Control because it's still updated. (#546)

futtta committed:
adding hook for full cache purges (#537)

* adding hook for full cache purges

* fire action outside the while loop as proposed by @donnchawp

donnchawp committed:
Remove and add spaces :) (#547)

donnchawp committed:
Update changelog and versions (#548)

Version 1.6.0

Commit:
Fix plugin name - WP Crontrol (#549)

* Fix plugin name - WP Crontrol

* Fix name of plugin - WP Crontrol

donnchawp committed:
Don't die() if there's a problem deleting files. wp_mail() instead. (#551)

* Don't die() if there's a problem deleting files. wp_mail() instead.

* Add the language domain to these translations

* Added filter "wpsc_send_uninstall_errors" to disable sending errors

Return 0 and uninstall errors won't send an email to the admin

donnchawp committed:
Add a notice when settings can't be updated. (#552)

Fixes #535 #550
The plugin would silently fail to update options since it writes to a
temporary file before copying over the config file. The config file is
writeable but the plugin won't have permission to create a new file if
the holding directory (usually wp-content) is not writeable.

donnchawp committed:
Use system temp folder and warn on rename error. props @arendvw (#553)

See #535
Use the system temporary folder to create the temporary configuration
file before overwriting the current file with rename().
Warn if the rename fails as settings won't be updated.

donnchawp committed:
Fix minor typo in this file. props @Chezus
Fixes #534

donnchawp committed:
Update version to 1.6.1 and changelog. (#554)

viniciusbarros committed:
Fix/avoid serving expired file (#562)

* Preventing serving expired cache, when directly serving file

* Preventing serving expired cache, when directly serving file

* Fix spaces in new code

Minor change: Fix indention of new code, and add spaces in code.

donnchawp committed:
Write directly to the config file to avoid permission issues (#563)

Fixes #520
A directory needs to be writeable to create a new file in it, but
modifying an existing file doesn't.

zytzagoo committed:
Fix #556 (#557)

Inverting the condition in order to actually use a proper path if it's not already set. Fixes #556

donnchawp committed:
Allow the Vary and Cache-Control headers to be modified. (#555)

* ALlow the Vary and Cache-Control headers to be modified.
Fixes #523 and #522
This is for advanced users only!

Define the constants WPSC_VARY_HEADER and WPSC_CACHE_CONTROL_HEADER to
override the default Vary and Cache-Control headers.
If the Vary header doesn't include Accept-Encoding visitors may get
mixed compressed or not pages.
If it is missing Cookie the wrong file might be sent to anonymous users.

* If headers are blank then don't add them to the cache/.htaccess

donnchawp committed:
Check if set_transient() is set before using it. (#565)

emrikol committed:
Fixes typo in error message. (#566)

I believe this is a typo, from what I can tell.  If not, please disregard :)

donnchawp committed:
Remove searchengine plugin (#567)

* Removed the searchenging example plugin

This plugin sets a cookie to track user interactions. It's only a test
plugin but raises privacy concerns so I feel it would be better to
remove it and replace it with a link to the file in the readme.txt

* Link to the searchengine.php on Github because it's removed here.

donnchawp committed:
Bump version to 1.6.2 and update changelog (#564)

* Bump version to 1.6.2 and update changelog

* Updated changelog

donnchawp committed:
Only check the cache max time if it's not 0, and GC disabled. (#571)

By setting the cache max time to 0 garbage collection is disabled but
this test would stop PHP serving cache files because it detected that
they were expired.
Fixes
https://wordpress.org/support/topic/version-1-6-2-fails-to-cache-had-to-revert-to-1-6-1/

donnchawp committed:
Don't clean expired files when GC is disabled. (#572)

This hopefully fixes the problems where cache files are deleted after
preload:
https://wordpress.org/support/topic/preload-cache-doesnt-seem-to-preload-all-pages/
https://wordpress.org/support/topic/issue-with-preloading/

donnchawp committed:
When deleting a single post, don't try to delete the homepage+sub dirs (#573)

* When deleting a single post, don't try to delete the homepage+sub dirs

Preloading deletes the old cache files for each page as it goes but
sometimes it deletes the homepage and then deletes everything else
because they're sub directories.
See:
https://wordpress.org/support/topic/preload-cache-doesnt-seem-to-preload-all-pages/

* If clear_post_supercache() tries to delete the homepage then stop

It should only delete posts or pages.

* Delete the homepage cache when preloading finishes.

donnchawp committed:
Add warning about DISABLE_WP_CRON (#575)

Fixes #532

stodorovic committed:
Fixes wp-cli - infinite amount of semaphores via sem_get (#576)

donnchawp committed:
Make it easier to add new WP Super Cache plugins by outside code. (#574)

* Put plugins in wp-content/plugins/wp-super-cache-plugins/

Adding new plugins to WP Super Cache to extend it is complicated but
where they should go should be easy. This patch makes the plugin check
in a particular directory if it exists and load the PHP files there.
The files there won't be deleted when WP Super Cache is updated.

* Add $wpsc_plugins array with plugins to be loaded

Add helper functions to add and delete plugin files from the list to be
loaded by WP Super Cache:
wpsc_add_plugin( $file );
wpsc_delete_plugin( $file );
And this function to return the list of plugins:
wpsc_get_plugins();

* Add documentation on adding/deleting plugins.

stodorovic committed:
Improves delete cache in admin bar (#578)

donnchawp committed:
Check the URL to avoid PHP warnings. ref: https://wordpress.org/support/topic/add-check-before-strpos/ (#579)

donnchawp committed:
Add cookie helper functions to modify the cache key more easily. (#580)

* Add cookie helper functions to modify the cache key more easily.

Inspired by
https://wordpress.org/support/topic/php-cookies-doesnt-work-with-wp-super-cache/
This will make it much easier for plugin authors that use cookies to
identify visitors to integrate with WP Super Cache.
Use the function wpsc_add_cookie( $name ) to add a cookie name to the
list of cookies and wpsc_delete_cookie( $name ) to remove it.

Example code to add the cookie "euCookie" used by the EU Cookie Law
plugin;
https://wordpress.org/plugins/eu-cookie-law/

--------------
function add_wpsc_cookies() {
    if ( function_exists( 'wpsc_add_cookie' ) ) {
        wpsc_add_cookie( 'euCookie' );
    }
}

add_action( 'init', 'add_wpsc_cookies' );
--------------

It's wrapped in an action on "init" to make sure that WP Super Cache is
loaded but the wpsc_add_cookie() function only needs to be called once.
Calling it multiple times does not add multiple entries for the same
cookie name.

* Add extra cookie names to mod_rewrite rules

* Add docs on plugin and cookie functions

* Extra documentation on cookie names+values

stodorovic committed:
Improve deleting of cache in edit/delete/publish actions - rev (#577)

stodorovic committed:
Optimize apache_response_headers (#496)

donnchawp committed:
Add wpsc_cookies and wpsc_plugins related actions to modify those settings (#582)

This PR adds four actions to modify wpsc_cookies and wpsc_plugins:

* wpsc_add_plugin
* wpsc_delete_plugin
* wpsc_add_cookie
* wpsc_delete_cookie

These actions will allow WordPress plugins to modify the cookies and
plugins lists used by WP Super Cache instead of the add/delete functions
introduced in #574 and #580.
Duplicate entries are removed. Example code:

For example, to add a cookie name called 'euCookie':

`do_action( 'wpsc_add_cookie', 'euCookie' );`

To remove that cookie:

`do_action( 'wpsc_delete_cookie', 'euCookie' );`

stodorovic committed:
Uses strpos instead of strncmp (#584)

Presskopp committed:
typo export mode -> expert mode (#585)

donnchawp committed:
Update version to 1.6.3 and changelog (#581)

* Update version and changelog

* Updated changelog

stodorovic committed:
Populates $WPSC_HTTP_HOST for WP-CLI commands (#587)

stodorovic committed:
Bump minimal requirement to WP 3.1 (#591)

stodorovic committed:
Fixes wpsc_post_transition and loads get_sample_permalink (#595)

stodorovic committed:
Populates globals in WP-CLI environment (#592)

stodorovic committed:
Fixes admin bar rendering (#589)

* Fixes admin bar rendering

* Uses "relative" path in admin_bar_delete_page and supercache_admin_bar_render

* Deprecation of favorite_actions

* Uses action admin_bar_menu

* Uses get_option( 'home' ) to remove base path

donnchawp committed:
Fix headings (#597)

* Make headings smaller.

Fixes https://wordpress.org/support/topic/wp-super-cache-admin-heading-structure-is-incorrect/

donnchawp committed:
Bump version number and add changelog (#598)

* Bump version numbe and add changelog

* Add another change

stodorovic committed:
Uses is_multisite() instead of constants (#600)

stodorovic committed:
Uses home_url() for admin bar (#603)

* Uses home_url() for admin bar

* Replaces wp_redirect with wp_safe_redirect for admin bar

* PHPCS fixes for admin_bar_delete_page

stodorovic committed:
Code refactoring inside wp-cache-base.php (#602)

stodorovic committed:
Uses get_post_status instead of post_status (#623)

stodorovic committed:
Fixes button - Update Direct Pages (#622)

* Fixes button - Update Direct Pages

* Fixes buttons in function wp_lock_down

stodorovic committed:
Removes apache_response_headers and uses only headers_list (#618)

stodorovic committed:
is_site_admin has been deprecated (#611)

stodorovic committed:
Fixes action urls in wp_cache_manager (#610)

donnchawp committed:
Remove the link to the HibbsLupusTrust tweet. (#635)

They're not using WP Super Cache any more.
https://twitter.com/HibbsLupusTrust/status/136429993059291136

stodorovic committed:
Don't load wp-cache-config.php if it's already loaded (#605)

stodorovic committed:
PHPCS fixes and optimization for plugins/domain-mapping.php (#615)

* PHPCS fixes and optimization for plugins/domain-mapping.php

* Uses wpsc_delete_files/wpsc_rebuild_files in plugins/domain-mapping.php

* Revert conditionals for function_exists

stodorovic committed:
Introduces PHP_VERSION_ID for faster checking (#604)

* Introduces PHP_VERSION_ID for faster checking

* Correction for ini_get( 'safe_mode' )

stodorovic committed:
Fixes regex and optimizes ossdl-cdn.php (#596)

* Fixes regexp and optimizes ossdl-cdn.php

* Fixes option - ossdl_off_cdn_url

* PHPCS fixes in ossdl-cdn.php

* CDN - improves filter and PHPCS/PHPMD fixes

donnchawp committed:
Bump version tested to 5.0.2 (#639)

donnchawp committed:
Promote the Jetpack Site Accelerator on the CDN page. (#636)

* Promote the Jetpack Site Accelerator on the CDN page.

The free site accelerator is a part of Jetpack which is commonly
installed on WordPress blogs so we should let users know about it.

* Disable CDN if SA enabled and update text descriptions.

* Add link to install Jetpack and make the "not found" text strong.

donnchawp committed:
Only update new settings and use a temporary file to avoid corruption. (#652)

* Only update new settings and use a temporary file to avoid corruption.
Fixes #648
The configuration file can be corrupted if multiple writers try to
update it at once. This patch updates a temporary file that is renamed
to the original name which should fix this.
It also limits updates to settings that have been updated. This should
decrease updates to the file by a huge amount.

* Minor changes to debug lines to indicate function.

donnchawp committed:
Serve cached files to rejected user agents, don't cache them. (#658)

Fixed https://wordpress.org/support/topic/caching-working-but-pingdom-and-bots-not-getting-supercache-files/#post-11293649

stodorovic committed:
Combine multiple headers with the same name (#641)

shanomurphy committed:
Open ‘Delete Cache’ link in same window (#656)

Open the ‘Delete Cache’ link in the admin bar in the same window instead of a new window.

donnchawp committed:
Check advanced-cache.php for WP Super Cache strings before modifying it. (#666)

* Check advanced-cache.php for WP Super Cache strings before modifying it.

If the file exists and doesn't have WP Super Cache strings then complain
and stop working.
Don't delete it if it doesn't have those strings either.

* Warn the user they must fix advanced-cache.php before using plugin.

* Check for plugin strings here before stopping an install of the plugin.

The plugin may have been deactivated incorrectly and the
advanced-cache.php left over. If so we should just use it again rather
than give out to the user.

donnchawp committed:
When updating settings always update blank lines. (#667)

If a new line is blank we don't want to test it against the current
line because a blank link will match the new line. This happens when
uninstalling or deactivating the plugin and removing WP_CACHE or
WPCACHEHOME from wp-config.php

donnchawp committed:
Bump version to 1.6.5 and add changelog (#669)

donnchawp committed:
Fix updating wpcachehome (#670)

* Set a transient "setting not changed" when settings not changed.

When modifying wp-config.php to add WPCACHEHOME check that transient as
the replace line function will return false since the line may not have
changed..

* Check if setting not changed when updating WP_CACHE too.

* Remove this line as it's completely incorrect.

For more:
https://wordpress.org/support/topic/instruction-to-modify-cachehome-is-not-clear/

* wp_cache_replace_line returns true no change, not false

It used to return false on fatal errors like permission or file missing
problems as well as if setting hadn't changed. This patch makes it
return true if setting hasn't changed.
I had to change other code to cope with that and reversed other changes
in this patch.

donnchawp committed:
Rename the temp config file correctly and remove wp-admin/.php files (#673)

The rename created a file called ".php" in wp-admin before writing to
the configuration file. The file is empty but this patch adds a fixer
function to delete it and also fixes the rename.

Reported:
https://wordpress.org/support/topic/php-notice-undefined-variable-13/
https://wordpress.org/support/topic/wp-super-cache-created-a-stray-php/

donnchawp committed:
Check if WPCACHEHOME exists in the global config file before adding it. (#674)

This fixes a problem adding the advanced-cache.php where it fails if the
file wp-config.php is read only. It now checks if WPCACHEHOME is in the
config file before checking if it's readonly.
ref: https://wordpress.org/support/topic/wpcachehome-must-be-set-in-config-file/

donnchawp committed:
Prepare for version 1.6.5 (#675)

Update versions, bump tested version, and add changelog.

donnchawp committed:
Save bools through the wp_cache_setting function. (#676)

donnchawp committed:
Check if $super_cache_enabled is true this way (#677)

In 1.6.5 $super_cache_enabled was saved as 1 rather than true. This
check failed so I'm changing it back to "==" and also fixed saving of
$super_cache_enabled in #676 because I don't want users to toggle
caching just to use preload again.

donnchawp committed:
Bump version to 1.6.7 and add changelog (#678)

donnchawp committed:
Add wpsc_is_caching_user_disabled to #631 (#657)

* Improves logic for logged in users

* Make the "not caching" option recognise logged in and cookie users.

Add the wpsc_is_caching_user_disabled() function that uses the
$wp_cache_not_logged_in setting to check the auth cookies or
all cookies to determine if caching should be disabled for the
current visitor.

* Add html comment for "caching disabled" messages.

* Use wp_footer instead of register_shutdown_function. Props @stodorovic

* Change language of cache restrictions section of settings page.

* Change spaces to TABs.

* REST API: update wp_cache_not_logged_in correctly

Add support for value == 0, 1, 2 instead of just 0 and 1.
0 - cache pages for all users.
1 - disable caching for users with cookies
2 - disable caching for logged in users

donnchawp committed:
Remove the object cache support. It never worked properly. (#681)

Sites are better off using Batcache or some other memory based caching
solution.

donnchawp committed:
Also check if WPCACHEHOME is set correctly when checking it. (#683)

It's not enough to check if WPCACHEHOME is in wp-config.php. I must also
check if it's either blank or set to the wrong path. If so then the
config file has to be updated.
Ref:
https://wordpress.org/support/topic/warning-wp-super-cache-caching-was-broken-but-has-been-fixed-3/

donnchawp committed:
Add new constants to change hos caching works.
Modification of #672 and fixes #682
WPSC_SERVE_DISABLED - if defined will disable serving any cache file.
WPSC_SUPERCACHE_ONLY - if defined only supercache fiels will be used.

donnchawp committed:
Hide get_post() warnings if no post found. (#684)

When deleting pages related to a post being deleted you need to get the post. If it has already been deleted then it does nothing. Some sites generate PHP warnings so this hides that.

Ref: https://wordpress.org/support/topic/some-errors-19/

donnchawp committed:
Update docs and versions (#685)

donnchawp committed:
Improve some advanced-cache.php variables and messages (#687)

Rename $wp_cache_link to $wpsc_advanced_cache_filename
Rename $wp_cache_file to $wpsc_advanced_cache_dist_filename
Rename wp_cache_check_link() to wpsc_check_advanced_cache()
Sometimes the advanced-cache.php is created by another plugin or the
hosting provider and should not be overwritten by this plugin.
Added a message to contact the hosting support if the error keeps
reappearing.

donnchawp committed:
Add a warning to the log file viewer (#688)

Two many people expose sensitive information about their sites..

donnchawp committed:
Only allow the debug log to be viewed through the viewer. (#691)

donnchawp committed:
Debug log cleanup (#692)

* Debug log cleanup. remove cookies and browsers, escape some characters.

Don't show cookies and browser UA in the debug log. Avoids data leaking
if the user posts the log publicly.
Run each line through htmlspecialchars() to be careful.

* Debug log viewer excludes wp-admin and json by default.

The user can uncheck those settings if they want.

donnchawp committed:
Some very minor whitespace changes. (#693)

donnchawp committed:
Replace the ABSPATH in the debug log to avoid leaking sensitive data (#694)

donnchawp committed:
Added wpsc_update_check() to to run on plugin updates.

This code will delete the debug log when the plugin is updated to
version 1.6.9.

donnchawp committed:
Getting ready for 1.6.9 (#695)

* Getting ready for 1.6.9

* Update the changelog

donnchawp committed:
Added "wpsc_cdn_urls" filter to modify the URLs used to rewrite URLs (#697)

This filter will allow other plugins to modify the URLs used to rewrite
URLs in the html by the CDN feature.

Prompted by
https://wordpress.org/support/topic/wp-super-cache-doesnt-offer-cdn-support-for-multiple-domain-wpml-installations/

donnchawp committed:
Move these checks below OB so CDN works again. (#698)

Those checks stopped the output buffer working which stopped the CDN
function working for logged in/known users when caching was disabled for
them.
Moving them down here allows the OB to work and the CDN code runs too.

donnchawp committed:
Make it clear which settings only work in simple mode. (#699)

Some of the settings don't work in expert mode because files are served
by mod_rewrite rules so make it clear which ones are affected by
disabling them and printing a message. Prompted by:
https://wordpress.org/support/topic/304-not-modified-greyed-out/

donnchawp committed:
Don't enable mobile support by default. (#700)

Mobile support is too complicated in this plugin and is probably not
required as themes resize to suit the browser. It can still be enabled
from the advanced settings page.

Commit:
Updated 'admin bar' verbiage to 'Toolbar' to match the change in verbiage found in core WP. (#701)

donnchawp committed:
Added summary of "easy" settings applied. (#703)

donnchawp committed:
Release 170 (#704)

* Update version number and changelog.

* Update changelog

donnchawp committed:
Add debugging to these deletes (#705)

donnchawp committed:
Bump the tested version to 5.3.2 (#723)

donnchawp committed:
Update tested 532 (#724)

* Bump the tested version to 5.3.2

* Check if $value is an array to avoid PHP warnings.

Props @chillinchiz
ref: https://wordpress.org/support/topic/php-7-4-and-wp-supercache/#post-12335149

jakejarvis committed:
Improve wording of file list headers (#719)

raphj committed:
Fix it's -> its and document wp-super-cache-plugins (#709)

* Fix it's -> its

* Document the wp-super-cache-plugins directory

bonny committed:
Update origin pull link (#645)

* Update origin pull link

It was going to a 404, probably because SoftLayer was acquired by IBM and is now called/part of IBM Cloud.

* Update ossdl-cdn.php

Change origin pull link again, to a Google search.

Co-authored-by: Donncha Ó Caoimh <donncha@automattic.com>

donnchawp committed:
Add a link to crowdsignal.com in the settings page. (#725)

My day job, we do a pretty good job od serving polls and surveys!

donnchawp committed:
Getting ready for version 1.7.1 (#726)

Bump version number, changelog, updated contributors list.

donnchawp committed:
is_writeable works fine on non Windows systems. (#742)

* is_writeable works fine on non Windows systems.

* Check if ! WIN after checking for Darwin.

Match a whole lot more Unix systems.

* Yoda this.

donnchawp committed:
Add more advanced-cache checks and better messaging. (#741)

Some hosts don't want the advanced-cache.php to be overwritten so we
should handle that. If another caching plugin was installed make sure
that we don't clobber their file.
Don't delete the advanced-cache.php if it doesn't belong to this plugin.

donnchawp committed:
Fixed an undefined variable warning.
Fixes #739

smortex committed:
Fix plugin for usage with Debian wordpress package (#730)

* Fix plugin for usage with Debian wordpress package

Debian packages wordpress in a way that a single wordpress instance can
be used to serve multiple websites.  The wp-config.php file is stored
in a well-known location: /etc/wordpress/config-$HOST.php as explained
in the [README.Debian] file.

The package defines DEBIAN_FILE as the filename of the configuration
file.  When it is defined, and if the wp-config.php was not found using
the "traditional" way, use it.

By default, this file is not writable by the application, adding the
following to it is enough to use the plugin:

define('WP_CACHE', true);
define('WPCACHEHOME', '/var/lib/wordpress/wp-content-xxxxx/plugins/wp-super-cache/');

[README.Debian]:https://salsa.debian.org/debian/wordpress/blob/master/debian/README.debian

* Fix formatting of patch

* Check if this file exists

We're checking other locations, might as well check here too.

Co-authored-by: Donncha Ó Caoimh <donncha@automattic.com>

donnchawp committed:
Added Docker compose file to create docker container. (#745) (#746)

donnchawp committed:
Initial import of framework for rewrite.

Also includes some work on configuration saving.

donnchawp committed:
Use the right text domain

donnchawp committed:
Add $filename to replace_line_in_file() so it's more useful.

We'll use it to write to the global config file.

donnchawp committed:
Add functions to check setup.

The functions will check if advanced-cache.php exists and WP_CACHE is
defined in wp-config.php

donnchawp committed:
In admin page check the setup before showing settings.

The plugin will attempt to create the file advanced-cache.php  and add
WP_CACHE to wp-config.php

CGastrell commented:
Since you're working almost "solo" on this, I'd just approve it. But if you come up with some simple tests so I can say "works", I'd have no problem testing it :)

donnchawp committed:
Remove ", so the path to the plugin is correct

donnchawp committed:
Check for WPCACHEHOME and improve advanced-cache.php creation

Use a HEREDOC to create the advanced-cache.php and add functions to
check and add WPCACHEHOME to wp-config.php

donnchawp committed:
Add wp_super_cache_create_cache() to process output buffer.

Rename the get_cache_dir() function so it has the right prefix.

donnchawp committed:
Fill in more code from the plugin.

* Use the config class to get the configuration as we need it to write
too for debugging.
* Add a new class to add debugging.

Next TODO: create a blank configuration files to save to.

donnchawp committed:
Add code to create a blank configuration file.

donnchawp committed:
Add a "$caching = 0" to the config file.

donnchawp committed:
Change caching to cache_enabled

donnchawp committed:
Add a User class and move code in there.

donnchawp committed:
$config is an object, not an array.

donnchawp committed:
Ignore this, the plugin does lots of filesystem actions.

donnchawp committed:
Add sample config file

donnchawp committed:
Fix case of config class and add wp_cache_home_path setting function.

donnchawp committed:
Change how the $config is used.

Instead of creating a local $config just use $this->config->config.

donnchawp committed:
Lots of changes:

* Renamed caching to file-cache
* Add cache debug setting to admin page.
* Added Page class to represent the current page.
* And more ..

donnchawp committed:
Use the gzip_encoding() function

donnchawp committed:
Check "cache_compression" setting in the gzip_encoding function

donnchawp committed:
Rename function ok_to_cache to is_cacheable

donnchawp committed:
Check if this exists before setting it.

stathis commented:
While you are at it, why don't you add php-brotli support to it?
I'm already maintaining a fork of WPSC with brotli support, maybe it could be of help.

[This extension](https://github.com/kjdev/php-ext-brotli) is already in some repositories, and most Nginx nowadays support `brotli_static` .

donnchawp committed:
Add make_anonymous()

donnchawp committed:
Fixed up is_caching_disabled and debugging logs

donnchawp committed:
Move cache_init into Page class contructor

donnchawp committed:
Minor changes, for phpcs

donnchawp committed:
Fix "cache_enabled" check and add customizer check

donnchawp committed:
Fix call to gzip_encoding()

donnchawp committed:
Cache query_vars to speed it up.

donnchawp committed:
Split cache checks into pre and post.

Some checks on "cacheability" can only be run after WordPress has loaded
so query vars are defined.

donnchawp committed:
Remove wp_super_cache_config and clean up debug code.

donnchawp committed:
Forgot a description for this function

donnchawp committed:
phpcs fix and move post checks into output buffer handler

donnchawp committed:
Oops, this is an array, not a function.

donnchawp committed:
Move post check into this output buffer function

donnchawp committed:
Multiple fixes.

Fix UA check, remove url check, and fix privacy status of post check
function.

donnchawp committed:
Move $wp_query check to get_query_vars()

donnchawp committed:
Tiny typo in 404 query var.

donnchawp committed:
Added set_env function.

This function will cache various blog settings for the shutdown process
since $wpdb may have been destroyed by the time shutdown happens.

donnchawp committed:
Multiple changes.

* Move query vars into config class.
* Add get_cache_meta_information() for non-anon requests.
* Add Page set_env() to replace base file to setup request info.
* Replace wp_super_cache_request_uri with WPSC_URI constant

donnchawp committed:
Multiple changes:

* Add get_response_headers
* Add get_post_id
* Fire wp_set_env on 'template_redirect' so query vars are set.
* Put post_id in a config variable.

donnchawp committed:
Add functions to create cache folders and call from setup.

donnchawp committed:
Add page to the admin so environment is setup.

* Fix the blog cache path since only MU installs with have a "blogs"
directory.

donnchawp committed:
Move cache directory creation to file cache and call via filter.

This will allow us to use different cache storage backends in the
future.

donnchawp committed:
This config variable isn't used any more.

donnchawp committed:
Multiple changes:

* Move DISABLE_SUPERCACHE and WPCACHEHOME checks into page class.
* Delete wp-cache-base code as it was moved elsewhere already.
* Define $wp_super_cache_config in pre-wp-cache.php and use object
rather than config array. Fix associated code.
* Delete wpsc_is_backend - it's moved elsewhere.

donnchawp committed:
Remove broken "cache_enabled" check and move debugging around.

donnchawp committed:
Remove this, not needed now.

donnchawp committed:
Tiny whitespace diff

donnchawp committed:
Get debugging working.

* Disable debugging while creating the debug log, or we run into a loop.
